proxy.ts

```diff
@@ -0,0 +1,107 @@
+/**
+ * @internal Symbol used to track if the object has been mutated
+ */
+export const _MUTATED = Symbol("mutated");
+
+/**
+ * @internal Symbol used to track if the object is being tracked
+ */
+export const _TRACKED = Symbol("tracked");
+
+type Mutatable<T> = T & { [_MUTATED]?: boolean; [_TRACKED]?: boolean };
+
+// for Set, Map, and Array
+const MUTATION_METHODS = new Set([
+    "set",
+    "add",
+    "delete",
+    "clear",
+    "push",
+    "pop",
+    "shift",
+    "unshift",
+    "splice",
+    "sort",
+    "reverse",
+    "fill",
+    "copyWithin",
+]);
+
+/**
+ * @internal Creates a proxy wrapper around an object to track mutations
+ * @param object - The target object to proxify
+ * @returns Proxied object that tracks mutations
+ */
+export function trackMutations<T>(
+    object: T,
+): Mutatable<T> {
+    if (typeof object !== "object" || object === null) {
+        return object as Mutatable<T>;
+    }
+
+    const proxyCache = new WeakMap();
+
+    let hasChanges = false;
+
+    // deno-lint-ignore no-explicit-any
+    const handler: ProxyHandler<any> = {
+        get(target, prop) {
+            if (prop === _TRACKED) return true;
+            if (prop === _MUTATED) return hasChanges;
+            const originalValue = Reflect.get(target, prop);
+            if (typeof originalValue === "object" && originalValue !== null) {
+                if (proxyCache.has(originalValue)) {
```

Comment:
Can we skip further proxification if `hasChanges` is already `true`? It might be a bit of a micro-optimisation so I would understand if you dislike this

Reply:
Good idea!

Reply:
Fixed in 21cf040, resolving